images/dnsmasq/validate_dynamic_deps.sh

```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
+# This script will export the files from the given image. It will then use
+# objdump to read dnsmasq's shared object dependencies and verify that they
+# are included in the image.
+#
+# Example:
+# ./verify_dynamic_deps.sh gcr.io/k8s-staging-dns/k8s-dns-dnsmasq-arm64:1.22.5 /usr/sbin/dnsmasq
+
+TARGET_IMAGE=$1
+CONTAINER_NAME=$(sed -r --expression "s%[^/]+/([^/]+)/([^:]+):(.*)%\2%" <<< "$TARGET_IMAGE")
+TAG=$(sed -r --expression "s%[^\/]+/([^\/]+)/([^:]+):(.*)%\3%" <<< "$TARGET_IMAGE")
+
+IMAGE_CONTENT_DIR="test-images/$CONTAINER_NAME-$TAG"
+mkdir -p "$IMAGE_CONTENT_DIR"
+(cd "$IMAGE_CONTENT_DIR" &&
+docker create --name "$CONTAINER_NAME" $TARGET_IMAGE > /dev/null && \
+(docker export $CONTAINER_NAME | tar x) && docker rm $CONTAINER_NAME >/dev/null)
+
+find_arch() {
+  echo "$(file "$1" | awk -F',' '{print $2}')"
+}
+
+BIN_PATH=$2
+
+BIN_LOCATION="$IMAGE_CONTENT_DIR/$BIN_PATH"
+if [ ! -f "$BIN_LOCATION" ]; then
+  echo "$BIN_LOCATION does not exist"
+  exit 1
+fi
+
+BIN_ARCH=$(find_arch "$BIN_LOCATION")
+# Find all of the required dependencies
+dependencies=($(objdump -p "$BIN_LOCATION" | awk '/NEEDED/{ print $2 }'))
+
+
+echo "Finding dependencies for $BIN_PATH $BIN_ARCH"
+
+missing=(  )
+declare -A seen wrong_arch
+
+while [ ${#dependencies[@]} -gt 0 ]; do
+  current=${dependencies[0]}
+
+  # Check if we have evaluated this dependency
+  if [[ -v 'seen[$current]' ]]; then
+    dependencies=( "${dependencies[@]:1}" )
+    continue
+  fi
+
+  current_location=$(find "$IMAGE_CONTENT_DIR" -name "$current" | head -n 1)
+
+  # This there were no files with the given name
+  if [[ "$current_location" == "" ]]; then
+    seen[$current]=true
+    missing+=( "$current" )
+    # Remove first element
+    dependencies=( "${dependencies[@]:1}" )
+    continue
+  fi
+
+  # Expand symlinks because objdump doesn't follow them
+  while [ -L "$current_location" ]; do
```

Comment:
TBH Would it be a good idea to have a python script instead of bash? It a little effort but that would be a major readability/supportability boost.

E.g. chain expand symlinks in python would be just: `os.path.realpath(current_location)`

Reply:
Yep, I'll rewrite this in python real quick.